### PR TITLE
Add logger extensions with event ids

### PR DESCRIPTION
**Pull request by jbagga**
Addresses #393

### CODE REVIEW

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/DatabaseErrorPageMiddleware.cs [hunk 1]
```diff
@@ -115,15 +116,15 @@ public void Hold(Exception exception, Type contextType)
 
                         if (context == null)
                         {
-                            _logger.LogError(Strings.FormatDatabaseErrorPageMiddleware_ContextNotRegistered(contextType.FullName));
+                            _logger.ContextNotRegistereDatabaseErrorPageMiddleware(contextType.FullName);
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/DatabaseErrorPageMiddleware.cs]: Missed the `d` in `Registered`

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+            _invalidContextType = LoggerMessage.Define<Type>(
+                LogLevel.Error,
+                2,
+                "The context type '{ContextType}' could not be loaded. Ensure this is the correct type name for the context you are trying to apply migrations for.");
+
+            _contextNotRegistered = LoggerMessage.Define<string>(
+                LogLevel.Error,
+                3,
+                "The context type '{Context}' was not found in services. This usually means the context was not registered in services during startup. You probably want to call AddScoped<>() inside the UseServices(...) call in your application startup code.");
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs]: `'{ContextType}'`

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+            _requestPathMatched = LoggerMessage.Define<string>(
+                LogLevel.Debug,
+                4,
+                "Request path matched the path configured for this migrations endpoint({RequestPath}). Attempting to process the migrations request.");
+
+            _applyingMigrations = LoggerMessage.Define<string>(
+                LogLevel.Debug,
+                5,
+                "Request is valid, applying migrations for context '{Context}'");
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs]: If this is the same value as `ContextType` above then it should be named `ContextType`

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+            _migrationsApplied = LoggerMessage.Define<string>(
+                LogLevel.Debug,
+                6,
+                "Migrations successfully applied for context '{Context}'.");
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs]: Same

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+            _migrationsEndPointMiddlewareException = LoggerMessage.Define<string, Exception>(
+                LogLevel.Error,
+                7,
+                "An error occurred while applying the migrations for '{Context}'. See InnerException for details.: {Exception}");
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs]: Remove the period

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/MigrationsEndPointMiddleware.cs [hunk 6]
```diff
@@ -120,7 +121,7 @@ public class MigrationsEndPointMiddleware
             {
                 var message = Strings.FormatMigrationsEndPointMiddleware_InvalidContextType(contextTypeName);
 
-                logger.LogError(message);
+                logger.InvalidContextType(contextType);
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/MigrationsEndPointMiddleware.cs]: Should use the `contextTypeName`

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+            _matched = LoggerMessage.Define(
+                LogLevel.Debug,
+                4,
+                "Entity Framework recorded that the current exception was due to a failed database operation. Attempting to show database error page.");
+
+            _contextNotRegisteredDatabaseErrorPageMiddleware = LoggerMessage.Define<string>(
+                LogLevel.Error,
+                5,
+                "The context type '{Context}' was not found in services. This usually means the context was not registered in services during startup. You probably want to call AddScoped<>() inside the UseServices(...) call in your application startup code. Skipping display of the database error page.");
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs]: `ContextType`

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+            _notRelationalDatabase = LoggerMessage.Define(
+                LogLevel.Debug,
+                6,
+                "The target data store is not a relational database. Skipping the database error page.");
+
+            _databaseErrorPageMiddlewareException = LoggerMessage.Define<Exception>(
+                LogLevel.Error,
+                7,
+                "An exception occurred while calculating the database error page content. Skipping display of the database error page.: {Exception}");
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs]: Remove the exception part and pass the exception normally.

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Internal
+{
+    internal static class DiagnosticsEntityFrameworkCoreLoggerExtensions
+    {
+        private static readonly Action<ILogger, Exception> _noContextType;
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs]: I prefer this method for defining these: https://github.com/aspnet/KestrelHttpServer/blob/6584a8b5fdaba0a79c52b981cac88472bcc92d1a/src/Kestrel.Core/Internal/Infrastructure/KestrelTrace.cs#L12

No need for a static constructor

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+        private static readonly Action<ILogger, string, Exception> _contextNotRegisteredDatabaseErrorPageMiddleware;
+        private static readonly Action<ILogger, Exception> _notRelationalDatabase;
+        private static readonly Action<ILogger, Exception> _noRecordedException;
+        private static readonly Action<ILogger, Exception> _noMatch;
+        private static readonly Action<ILogger, Exception> _matched;
+        private static readonly Action<ILogger, Exception, Exception> _databaseErrorPageMiddlewareException;
+
+        static DiagnosticsEntityFrameworkCoreLoggerExtensions()
+        {
+            _noContextType = LoggerMessage.Define(
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs]: I've been putting a comment at each section of log messages with the logger category name to make it obvious why the event id's reset etc.

File: src/Microsoft.AspNetCore.Diagnostics/Internal/DiagnosticsLoggerExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+                "The response has already started, the error handler will not be executed.");
+
+            _errorHandlerException = LoggerMessage.Define<Exception>(
+                LogLevel.Error,
+                3,
+                "An exception was thrown attempting to execute the error handler: {Exception}");
+
+            _responseStartedErrorPageMiddleware = LoggerMessage.Define(
+                LogLevel.Warning,
+                2,
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Diagnostics/Internal/DiagnosticsLoggerExtensions.cs]: event id seems wrong

File: src/Microsoft.AspNetCore.Diagnostics/Internal/DiagnosticsLoggerExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+                "The response has already started, the error page middleware will not be executed.");
+
+            _displayErrorPageException = LoggerMessage.Define<Exception>(
+                LogLevel.Error,
+                3,
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Diagnostics/Internal/DiagnosticsLoggerExtensions.cs]: same

File: src/Microsoft.AspNetCore.Diagnostics/Internal/DiagnosticsLoggerExtensions.cs (hunk 11 above)
Comment (jbagga) [reply src/Microsoft.AspNetCore.Diagnostics/Internal/DiagnosticsLoggerExtensions.cs]: It's from a different logger that uses the same logger message with event id 1 but different ones after it. I should add a comment to make that clear

Comment (BrennanConroy) [reply src/Microsoft.AspNetCore.Diagnostics/Internal/DiagnosticsLoggerExtensions.cs]: Here's what we have in SignalR for that case
https://github.com/aspnet/SignalR/blob/11cc57ee0e4ebc38c2cf5877b1df19e4e686f446/src/Microsoft.AspNetCore.SignalR.Client.Core/Internal/SignalRClientLoggerExtensions.cs#L82-L97

**Comment by anurse** [comment]
Can we put names in our EventIds as well? The `EventId` type takes a name parameter as well as just the integer. EF does this and it's way easier to query in structured logging sources that provide it. It avoids needing to look up obscure integers.

So:

```
_migrationsApplied = LoggerMessage.Define<string>(
  LogLevel.Debug,
  new EventId(6, "[Some Prefix].MigrationsApplied"),
  "Migrations successfully applied for context '{Context}'.");
```

**Comment by jbagga** [comment]
> put names in our EventIds as well

I think it is a good idea. We should consider that for MVC too.

 cc @Eilon

**Comment by anurse** [comment]
I've got a boilerplate issue I've been filing around the repos, I "copied" it to MVC ;) https://github.com/aspnet/Mvc/issues/6818

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs (hunk 3 above)
Comment (jbagga) [reply src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs]: All of these are actually `ContextType.FullName`, so using `ContextTypeName` instead. Thanks

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs (hunk 2 above)
Comment (jbagga) [reply src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs]: cc @Eilon These error messages are the ones I was referring to. Maybe they can be reworded

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs (hunk 7 above)
Comment (jbagga) [reply src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs]: cc @Eilon Another instance

**Comment by jbagga** [comment]
@BrennanConroy Updated

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs [hunk 13]
```diff
@@ -8,127 +8,111 @@ namespace Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Internal
...
+
+        // DatabaseErrorPageMiddleware
+        private static readonly Action<ILogger, Type, Exception> _attemptingToMatchException = LoggerMessage.Define<Type>(
+            LogLevel.Debug,
+            new EventId(1, "AttemptingToMatchException"),
+            "{ExceptionType} occurred, checking if Entity Framework recorded this exception as resulting from a failed database operation.");
+
+        private static readonly Action<ILogger, Exception> _noRecordedException = LoggerMessage.Define(
+            LogLevel.Debug,
+            new EventId(2, "AttemptingToMatchException"),
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs]: Different ID but same name as above?

**Comment by anurse** [comment]
Yay! Names! Thanks @jbagga !

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs (hunk 13 above)
Comment (jbagga) [reply src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs]: Copy paste error >.< Thank you!

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+            "Request is valid, applying migrations for context '{ContextTypeName}'");
+
+        private static readonly Action<ILogger, string, Exception> _migrationsApplied = LoggerMessage.Define<string>(
+            LogLevel.Debug,
+            new EventId(6, "MigrationsApplied"),
+            "Migrations successfully applied for context '{ContextTypeName}'.");
+
+        private static readonly Action<ILogger, string, Exception> _migrationsEndPointMiddlewareException = LoggerMessage.Define<string>(
+            LogLevel.Error,
+            new EventId(7, "Exception"),
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs]: "Exception" doesn't seem like a great name

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+            "The context type '{ContextTypeName}' was not found in services. This usually means the context was not registered in services during startup. You probably want to call AddScoped<>() inside the UseServices(...) call in your application startup code. Skipping display of the database error page.");
+
+        private static readonly Action<ILogger, Exception> _notRelationalDatabase = LoggerMessage.Define(
+            LogLevel.Debug,
+            new EventId(6, "NotRelationalDatabase"),
+            "The target data store is not a relational database. Skipping the database error page.");
+
+        private static readonly Action<ILogger, Exception> _databaseErrorPageMiddlewareException = LoggerMessage.Define(
+            LogLevel.Error,
+            new EventId(7, "Exception"),
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/Internal/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs]: Same

**Comment by jbagga** [comment]
@BrennanConroy Updated

File: test/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Tests/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Tests.csproj [hunk 16]
```diff
@@ -18,4 +18,8 @@
     <PackageReference Include="xunit.runner.visualstudio" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Service Include="{82a7f48d-3b50-4b1e-b82e-3ada8210c358}" />
```
Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Tests/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Tests.csproj]: Don't think you need this

File: test/Microsoft.AspNetCore.Diagnostics.Tests/Microsoft.AspNetCore.Diagnostics.Tests.csproj [hunk 17]
```diff
@@ -28,4 +28,8 @@
     <PackageReference Include="xunit.runner.visualstudio" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Service Include="{82a7f48d-3b50-4b1e-b82e-3ada8210c358}" />
```
Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.Diagnostics.Tests/Microsoft.AspNetCore.Diagnostics.Tests.csproj]: And this

File: test/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Tests/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Tests.csproj (hunk 16 above)
Comment (jbagga) [reply test/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Tests/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Tests.csproj]: VS adds these automatically

@Eilon are we ignoring these?

Comment (Eilon) [reply test/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Tests/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Tests.csproj]: If you remove it and VS keeps putting it back then keep it.

Comment (jbagga) [reply test/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Tests/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Tests.csproj]: Opening the solution in VS re-adds these. So checking them in

**Pull request merged**